core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -489,10 +495,6 @@ public void loadNext() throws IOException {
           }
           upstream = nextUpstream;
           nextUpstream = null;
-
-          assert(inMemSorter != null);
```

Comment:
What I meant to say is that my understanding of the old `free()` call here was performed when we hit the final record of the iterator, so I thought we'd still need / want to free the iterator at that point even in the new code.